spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -366,15 +370,55 @@ public void pruneColumns(StructType requestedSchema) {
 
   private Schema schemaWithMetadataColumns() {
     // metadata columns
-    List<Types.NestedField> fields =
+    List<Types.NestedField> metadataFields =
         metaColumns.stream()
             .distinct()
             .map(name -> MetadataColumns.metadataColumn(table, name))
             .collect(Collectors.toList());
-    Schema meta = new Schema(fields);
+    // only calculate potential column id collision if metadata column was requested
+    Schema metadataSchema =
+        metadataFields.isEmpty()
+            ? new Schema(metadataFields)
+            : deduplicateSchemaIds(metadataFields);
 
     // schema or rows returned by readers
-    return TypeUtil.join(schema, meta);
+    return TypeUtil.join(schema, metadataSchema);
+  }
+
+  private Schema deduplicateSchemaIds(List<Types.NestedField> metaColumnFields) {
+    Map<Integer, Types.NestedField> indexedMetadataColumnFields =
+        TypeUtil.indexById(Types.StructType.of(metaColumnFields));
+
+    // Calculate field ids to reassign from nested partition field
```

Comment:
I was initially leaning towards making this less impact (ie there may be some perf hit if the user requests _file and not _partition, what do you think?

If we go ahead with this approach maybe we should change  this comment though?
ie, 
// Calculate nested non-reserved metadata field ids to reassign

Reply:
good point, I remember that the `_file` was used a lot more than `_partition` so I added a check outside this function so that we will only apply dedup logic if `_partition` was explicitly requested.